R/1.0_module_data.R

```diff
@@ -0,0 +1,176 @@
+#' passed in multiple ways:
+#' 1. Directly as a [teal.data::teal_data()] object.
+#' 2. As a `reactive` object returning [teal.data::teal_data()]. [See section](#reactive-teal_data).
+#'
+#' @section Reactive `teal_data`:
+#'
+#' [teal.data::teal_data()] can change depending on the reactive context and `srv_teal` will rebuild
+#' the app accordingly. There are two ways of interacting with the data:
+#' 1. Using a `reactive` object passed from outside the `teal` application. In this case, reactivity
+#' is controlled by external module and `srv_teal` will trigger accordingly to the changes.
+#' 2. Using [teal_data_module()] which is embedded in the `teal` application and data can be
+#' resubmitted when needed by the user.
+#'
+#' Since server of [teal_data_module()] must return `reactive` `teal_data` object, it means that
+#' both scenarios (1) and (2) are having the same effect for the reactivity of a `teal` application.
+#' The difference is that in the first case the data is controlled from outside the app and in the
+#' second case the data is controlled from custom module called inside of the app.
+#'
+#' see [`validate_reactive_teal_data`] for more details.
+#'
+#' @inheritParams init
+#'
+#' @param data (`teal_data`, `teal_data_module` or `reactive` returning `teal_data`)
+#' @return A `reactiveVal` which is set to:
+#' - `teal_data` when the object is validated
+#' - `NULL` when not validated.
+#' Important: `srv_data` suppress validate messages and returns `NULL` so that `srv_teal` can
+#' stop the reactive cycle as `observeEvent` calls based on the data have `ignoreNULL = TRUE`.
+#'
+#' @rdname module_data
+#' @name module_data
+NULL
+
+#' @rdname module_data
+#' @keywords internal
+ui_data <- function(id, data, title, header, footer) {
+  ns <- shiny::NS(id)
+  shiny::div(
+    style = "display: inline-block;",
+    if (inherits(data, "teal_data_module")) {
+      actionButton(ns("open_teal_data_module"), NULL, icon = icon("database"))
+    } else {
+      NULL
+    }
+  )
+}
+
+#' @rdname module_data
+#' @keywords internal
+srv_data <- function(id, data, modules, filter = teal_slices()) {
+  checkmate::assert_character(id, max.len = 1, any.missing = FALSE)
+  checkmate::assert_multi_class(data, c("teal_data", "teal_data_module", "reactive", "reactiveVal"))
+  checkmate::assert_class(modules, "teal_modules")
+  checkmate::assert_class(filter, "teal_slices")
+
+  moduleServer(id, function(input, output, session) {
+    logger::log_trace("srv_data initializing.")
+
+    if (getOption("teal.show_js_log", default = FALSE)) {
+      shinyjs::showLog()
+    }
+
+    # data_rv contains teal_data object
+    # either passed to teal::init or returned from teal_data_module
+    data_validated <- if (inherits(data, "teal_data_module")) {
+      srv_teal_data_module(
+        "teal_data_module",
+        data = reactive(req(FALSE)), # to .fallback_on_failure to shiny.silent.error
```

Comment:
@gogonzo @m7pr @vedhav  this was changed after a call with @gogonzo today, but we should rethink and decide once and for all how do we deal with `shiny.silent.error`s

Line 79 was also changed to `req(FALSE)`

Options:

- Fallback to `NULL`
- Fallback to `shiny.silent.error`
- Don't fallback on initialization and just keep whatever error we get

Reply:
From a call this morning we settled on `shiny.silent.error` as it makes the most sense because it keeps an error state, while `NULL` isn't as explicit.

Validation error is shown via `{ui,srv}_validate_reactive_teal_data`, while data object passed around in teal application has a `shiny.silent.error`.

Does everyone agree?

Reply:
Sounds good. Glad we figured this out. I also got caught on this on Friday on another branch https://github.com/insightsengineering/teal/pull/1269/files#diff-bcbb361c0522bc4febd68d303caf283847bef473253a2734bdcb1bf68fd4eeeaR56-R68